### DISCUSSION TITLE
[Chapter 4] Q & A

### DISCUSSION THREAD

**Discussion by ahngilwoong** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#45] [CHAPTER 4 스트림 소개 - Q&A]
저희가 스트림을 왜 써야하는지에 대해서 배울 수 있는 시간이 되었는데
책 중간에 스트림은 매우 비싼 API라고 말한 내용이 있습니다.

### Q1) 스트림을 쓰면 안되는 경우???
단순히 매우 비싼데 왜 써? 부터 시작하였고 그럼에도 불구하고 장점이 많아서 쓴다! 라고 그 궁금증을 해결하였고
그러면 당연히 쓰면 안되는 상황도 있을 것 같아 다양한 의견을 듣고 싶어 질문합니다.

**Comment by coalong** [comment]
저도 책을 읽으면서 `스트림 api` 를 사용하지 않아야 될 상황은 언제일까? 생각해봤는데요.

p. 145  
### 4.3.1 스트림은 딱 한 번만 탐색할 수 있다. 
 
이 특징에 대해 생각해봤습니다.  스트림은 커다란 한 흐름으로 데이터를 처리합니다.
만약에 연산 중간중간에 값에 접근해야 될 경우에는, 
스트림을 최종 연산으로 끝내고 새로운 스트림을 생성하여 접근해야 된다고 알고 있습니다. 
이런 상황이라면 스트림 대신 컬렉션API 를 사용하여 데이터 처리를 하는 것이 더 좋다고 생각합니다.

**Comment by JoisFe** [comment]
예전에 길웅님께서 해당 주제로 따로 정리하신 글이 있지 않았나요? ㅎㅎ
링크 부탁드려요!
